enhancements/machine-api/control-plane.md

```diff
@@ -0,0 +1,339 @@
+
+### Goals
+
+- To have a declarative mechanism to manage the Control Plane as a single entity.
+- To support declarative safe horizontal scaling towards an odd number of replicas for the control plane compute resources.
+- To support declarative safe vertical scaling for the control plane compute resources.
+- To support declarative safe self-healing and replacement of Control Plane compute resources when the healthy resources are above n/2.
+- To support even spread of compute resources across multiple failure domains.
+
+### Non-Goals
+
+- To integrate with any existing etcd topology e.g external clusters. Stacked etcd with dynamic member identities, local storage and the Cluster etcd Operator are an assumed invariant.
+- To managed individual Control Plane components. Self hosted Control Plane components that are self managed by their operators is an assumed invariant:
+	- Rolling OS upgrades and config changes at the software layer are managed by the Machine Config Operator.
+	- etcd Guard that ensures a PDB to honour quorum is managed by the Machine Config Operator.
+	- Each individual Control Plane component i.e Kube API Server, Scheduler and controller manager are self hosted and managed by their operators.
+	- The Cluster etcd Operator manages certificates, report healthiness and add etcd members as "Master" nodes join the cluster.
+	- The Kubelet is tied to the OS and therefore is managed by the Machine Config Operator.
+- To integrate with any Control Plane components topology e.g Remote hosted pod based Control Plane.
+- Automated disaster recovery of a cluster that has lost quorum.
+- To manage OS upgrades. This is managed by the Machine Config Operator.
+- To manage configuration changes at the software layer. This is managed by the Machine Config Operator.
+- To manage the life cycle of Control Plane components
+- To automate the provisioning and decommission of the bootstrapping instance managed by the installer.
+
+## Proposal
+
+This proposes an ad-hoc CRD and controller for declaratively managing as a single entity the compute resources that host the OCP Control Plane components.
+
+This controller differs from a regular machineSet in that: 
+- It ensures that scaling operations are non disruptive for etcd:
+	- It scales one resource at a time.
+	- It let scaling operations proceed only when all etcd members are healthy and all the owned machines have a backed ready node.
+- It ensure even spread of compute resources across failure domains.
+- It removes etcd membership for voluntary machine disruptions (Question: can rather etcd operator somehow handle this?).
+- It owns and ensure safe values for a MHC that monitor the Control Plane machines.
+
+This entity is decoupled and orthogonal to the lifecycle and management of the Control Plane components that it hosts. All of these components are expected to keep self managing themselves as the cluster shrink and expand the Control Plane compute resources.
```

Comment:
Won't it lead to a flow where there are multiple managers to same component (and strong coupling between them). If it has to be aware of etcd orchestration logic - so why not to let the ceo to handle with it?

Reply:
The control plane components are already self managed by their operators orthogonally to the underlying compute resources. If we can assume ceo would react appropriately to "cattle" scale operations then we could use machineSets for masters. Resolving this to be discussed in a single thread -> https://github.com/openshift/enhancements/pull/278#discussion_r405695881